src/bijectors/stacked.jl

```diff
@@ -37,74 +37,65 @@ struct Stacked{Bs, N} <: Bijector{1} where N
         return new{A, N}(bs, ranges)
     end
 end
 Stacked(bs, ranges::AbstractArray) = Stacked(bs, tuple(ranges...))
 Stacked(bs) = Stacked(bs, tuple([i:i for i = 1:length(bs)]...))
 
-isclosedform(b::Stacked) = all(isclosedform.(b.bs))
+isclosedform(b::Stacked) = all(isclosedform, b.bs)
 
 stack(bs::Bijector{0}...) = Stacked(bs)
 
 # For some reason `inv.(sb.bs)` was unstable... This works though.
 inv(sb::Stacked) = Stacked(map(inv, sb.bs), sb.ranges)
-@generated function inv(sb::Stacked{A}) where {A <: Tuple}
-    exprs = []
-    for i = 1:length(A.parameters)
-        push!(exprs, :(inv(sb.bs[$i])))
-    end
-    :(Stacked(($(exprs...), ), sb.ranges))
-end
 
-# TODO: Is there a better approach to this?
 @generated function _transform(x, rs::NTuple{N, UnitRange{Int}}, bs::Bijector...) where N
     exprs = []
     for i = 1:N
         push!(exprs, :(bs[$i](x[rs[$i]])))
     end
-
     return :(vcat($(exprs...)))
 end
-_transform(x, rs::NTuple{1, UnitRange{Int}}, b::Bijector) = b(x)
-
+function _transform(x, rs::NTuple{1, UnitRange{Int}}, b::Bijector)
+    @assert rs[1] == 1:length(x)
+    return b(x)
+end
 function (sb::Stacked{<:Tuple})(x::AbstractVector{<:Real})
     y = _transform(x, sb.ranges, sb.bs...)
     @assert size(y) == size(x) "x is size $(size(x)) but y is $(size(y))"
     return y
 end
+# The Stacked{<:AbstractArray} version is not TrackedArray friendly
 function (sb::Stacked{<:AbstractArray, N})(x::AbstractVector{<:Real}) where {N}
-    y = vcat([sb.bs[i](x[sb.ranges[i]]) for i = 1:N]...)
+    init = vcat(sb.bs[1](x[sb.ranges[1]]))
+    y = mapreduce(vcat, 2:N; init = init) do i 
+        sb.bs[i](x[sb.ranges[i]])
+    end
     @assert size(y) == size(x) "x is size $(size(x)) but y is $(size(y))"
     return y
 end
 
 @views function (sb::Stacked)(x::AbstractMatrix{<:Real})
     init = reshape(sb(x[:, 1]), :, 1)
-    return mapreduce(i -> sb(x[:, i]), hcat, 2:size(x, 2); init = init)
-end
-
-@generated function logabsdetjac(
-    b::Stacked{<:Tuple, N},
-    x::AbstractVector{<:Real}
-) where {N}
-    exprs = []
-    for i = 1:N
-        push!(exprs, :(sum(logabsdetjac(b.bs[$i], x[b.ranges[$i]]))))
-    end
-
-    return :(sum([$(exprs...), ]))
+    return mapreduce(c -> sb(c), hcat, Base.Iterators.drop(eachcol(x), 1); init=init)
 end
 function logabsdetjac(
-    b::Stacked{<:AbstractArray, N},
+    b::Stacked{<:Any, N},
     x::AbstractVector{<:Real}
 ) where {N}
-    # TODO: drop the `sum` when we have dimensionality
-    return sum([sum(logabsdetjac(b.bs[i], x[b.ranges[i]])) for i = 1:N])
+    init = sum(logabsdetjac(b.bs[1], x[b.ranges[1]]))
+    init + sum(2:N) do i
+        sum(logabsdetjac(b.bs[i], x[b.ranges[i]]))
+    end
 end
+
 function logabsdetjac(b::Stacked, x::AbstractMatrix{<:Real})
-    return [logabsdetjac(b, x[:, i]) for i = 1:size(x, 2)]
+    @views init = vcat(logabsdetjac(b, x[:,1]))
+    return mapreduce(vcat, Base.Iterators.drop(eachcol(x), 1); init=init) do c
```

Comment:
`eachcol` isn't available in Julia-1.0

Reply:
I think it's defined in Compat. I can depend on that.